fixed-hash/src/hash.rs

```diff
@@ -321,6 +335,80 @@ macro_rules! construct_hash {
 	}
 }
 
+/// Implements conversion to and from hash types of different sizes. Uses the
+/// last bytes, e.g. `From<H256> for H160` uses bytes 12..32
+/// CAUTION: make sure to call with correct sizes and the bigger type first or
+/// bad things will happen!
+#[macro_export]
+macro_rules! impl_hash_conversions {
+	($a: ident, $a_size: expr, $b: ident, $b_size: expr) => {
+		impl From<$b> for $a {
+			fn from(value: $b) -> $a {
+				// REVIEW: better way of ensuring the macro params are ok?
+				assert!($a_size > $b_size && $a_size % 2 == 0 && $b_size %2 == 0);
+				let mut ret = $a::new();
+				ret.0[($a_size - $b_size)..$a_size].copy_from_slice(&value);
+				ret
+			}
+		}
+
+		impl From<$a> for $b {
+			fn from(value: $a) -> $b {
+				// REVIEW: better way of ensuring the macro params are ok?
+				assert!($a_size > $b_size && $a_size % 2 == 0 && $b_size %2 == 0);
+				let mut ret = $b::new();
+				ret.0.copy_from_slice(&value[($a_size - $b_size)..$a_size]);
+				ret
+			}
+		}
+
+		impl<'a> From<&'a $b> for $a {
+			fn from(value: &'a $b) -> $a {
+				let mut ret = $a::new();
+				ret.0[($a_size - $b_size)..$a_size].copy_from_slice(value);
+				ret
+			}
+		}
+	}
+}
+
+/// Implements conversion to and from a hash type and the equally sized unsigned int.
+/// CAUTION: Bad things will happen if the two types are not of the same size!
```

Comment:
maybe add a debug_assert along with caution?

Reply:
Sort of like I did [here](https://github.com/paritytech/rust-rocksdb)? What are your thoughts on `assert` vs `debug_assert` in this case?

Reply:
regular `assert` is not zero cost in production i believe

Reply:
Right, I was more asking if it's worth it to use `assert!` anyway or if the cost is prohibitive.

Reply:
what bad things ought to happen exactly? Isn’t this bad things already encompass panicking?

Reply:
@pepyakin you are correct! :) Added a `debug_assert`, which is similar to what `uint` does.